modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/openapi/contributor/FilterableFieldsOpenAPIContributor.java

```diff
@@ -0,0 +1,269 @@
+
+		if (xClassNameSchema == null) {
+			return null;
+		}
+
+		String xClassNameDefault = (String)xClassNameSchema.getDefault();
+
+		if (Validator.isBlank(xClassNameDefault)) {
+			return null;
+		}
+
+		Schema xSchemaNameSchema = properties.get("x-schema-name");
+
+		String xSchemaName = null;
+
+		if (xSchemaNameSchema != null) {
+			xSchemaName = (String)xSchemaNameSchema.getDefault();
+		}
+
+		List<EntityModelResource> entityModelResources =
+			_getEntityModelResources(
+				xClassNameDefault, jaxRsApplicationName, openAPIContext,
+				xSchemaName);
+
+		if (ListUtil.isEmpty(entityModelResources)) {
+			return null;
+		}
+
+		Map<String, EntityField> entityFieldsMap = new HashMap<>();
+
+		for (EntityModelResource entityModelResource : entityModelResources) {
+			MultivaluedHashMap<String, Object> params =
+				new MultivaluedHashMap<>();
+
+			params.putSingle("companyId", openAPIContext.getCompanyId());
+
+			EntityModel entityModel = entityModelResource.getEntityModel(
+				params);
+
+			if (entityModel == null) {
+				continue;
+			}
+
+			Map<String, EntityField> currentEntityFieldsMap =
+				entityModel.getEntityFieldsMap();
+
+			if (currentEntityFieldsMap != null) {
+				entityFieldsMap.putAll(currentEntityFieldsMap);
+			}
+		}
+
+		return entityFieldsMap;
+	}
+
+	private List<EntityModelResource> _getEntityModelResources(
+			String className, String jaxRsApplicationName,
+			OpenAPIContext openAPIContext, String schemaName)
+		throws Exception {
+
+		String filterString = null;
+
+		if (schemaName != null) {
+			filterString = StringBundler.concat(
+				"(entity.class.name=", className, "#",
+				StringUtil.toLowerCase(schemaName), ")");
+		}
+		else {
+			filterString = "(entity.class.name=" + className + ")";
+		}
+
+		ServiceReference<?>[] resourceServiceReferences =
+			_bundleContext.getServiceReferences(
+				(String)null,
+				StringBundler.concat(
+					"(&(api.version=",
+					GetterUtil.get(openAPIContext.getVersion(), "v1.0"),
+					")(osgi.jaxrs.resource=true)",
+					"(osgi.jaxrs.application.select=\\(osgi.jaxrs.name=",
+					jaxRsApplicationName, "\\))", filterString, ")"));
+
+		if (resourceServiceReferences == null) {
+			return null;
+		}
+
+		List<EntityModelResource> resources = new ArrayList<>();
+
+		for (ServiceReference<?> serviceReference : resourceServiceReferences) {
+			Object service = _bundleContext.getService(serviceReference);
+
+			if (service instanceof EntityModelResource) {
+				resources.add((EntityModelResource)service);
+			}
+		}
+
+		return resources;
+	}
+
+	private Set<String> _getFilterableFieldNames(
+		EntityField entityField, String fieldName,
+		Set<EntityField> processedEntityFields) {
+
+		if (!processedEntityFields.add(entityField)) {
+			return new HashSet<>();
+		}
+
+		if (!(entityField instanceof ComplexEntityField)) {
+			return SetUtil.fromArray(fieldName);
+		}
+
+		ComplexEntityField complexEntityField = (ComplexEntityField)entityField;
```

Comment:
I think Pablo told us to optimize to avoid duplicated fields in the list. Could we do it?